Consolidation of Martini Build CI/CD Implementation and Terraform Template

jasonquinto opened the pull request:
These changes are fairly extensive and focus on making the entire Martini Build CI/CD implementation easier to manage, review, and maintain. All Terraform modules, buildspecs, scripts, and supporting files are now consolidated into a single repository. This eliminates the drift and back-and-forth caused by the previous two-repo setup. The new structure streamlines versioning, keeps pipeline logic and infrastructure in sync, and provides contributors with one clear place to work from. Along the way, I aligned the documentation with the actual code, fixed inconsistencies, removed duplicates, and normalized patterns across both pipeline modules.

To improve long-term maintainability, I refactored the Terraform templates so we now have shared, reusable modules, clearer naming, and a cleaner variable structure. I also introduced local pre-commit hooks and Pre-Commit CI in GitHub Actions. This ensures that every change follows formatting, linting, and Terraform best practices before it reaches review, which saves reviewers time and helps catch issues early. Overall, the goal is to simplify and enforce consistency across all components of the CI/CD stack.

Highlights
- Consolidated all Terraform, buildspecs, Dockerfile, and scripts into one repo
- Added reusable shared modules to eliminate duplication
- Introduced local pre-commit + Pre-Commit CI for automated linting & formatting
- Fixed mismatches between READMEs and variables.tf (defaults, required flags, etc.)
- Cleaned up repeated variables across modules
- Standardized SSM usage, IAM, naming conventions, and CloudWatch log structure

Pre-commit CI result:
https://github.com/lontiplatform/martini-build-pipeline-aws/actions/runs/19657219732/job/56296237197

<img width="618" height="268" alt="Screenshot from 2025-11-25 11-36-03" src="https://github.com/user-attachments/assets/b56d34fd-534b-48ba-bc65-c81f013d496e" />

Commit:
Initial commit

Commit:
Revised README file

Commit:
Added requested information to README

jasonquinto committed:
Updated README file

jasonquinto committed:
Merge branch 'main' into development

Commit:
Improved client ID and secrets handling

Commit:
Updated the README file to reflect changes.

Commit:
Updated the README file to reflect changes.

jasonquinto committed:
Merge branch 'main' into development

jasonquinto committed:
Add support for long-lived oauth token and deprecated fetch method

jasonquinto committed:
Merge branch 'main' into development

jasonquinto committed:
Add allowed package feature

jasonquinto committed:
Merge branch 'main' into development

jasonquinto committed:
Updated the pipelines and dockerfile

jasonquinto committed:
Add fallbacks and URL sanitation

jasonquinto committed:
Updated the README to reflect changes.

jasonquinto committed:
Cleaned up the README

jasonquinto committed:
Removed AWS region and ID in martini-build-image, updated the README and improved the upload script

jasonquinto committed:
Merge branch 'main' into development

jasonquinto committed:
Swtich to ARM64 binaries

jasonquinto committed:
Merge branch 'development' of https://github.com/lontiplatform/martini-build-pipeline-aws into development

jasonquinto committed:
Added support details

jasonquinto committed:
Merge branch 'main' into development

jasonquinto committed:
Refactored and consolidated the implementation and terraform template

jasonquinto committed:
Fixing merge conflicts

jasonquinto committed:
Fix reference links

rafaljanicki commented on .github/workflows/precommit.yml:
```diff
@@ -0,0 +1,56 @@
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.11"
```
Let's use 3.13 or 3.14 if we're refactoring

rafaljanicki commented on .github/workflows/precommit.yml:
```diff
@@ -0,0 +1,56 @@
+      - name: Set up Terraform
+        uses: hashicorp/setup-terraform@v3
+        with:
+          terraform_version: "1.6.6"
```
Let's use Tofu instead

rafaljanicki commented on .github/workflows/precommit.yml:
```diff
@@ -0,0 +1,56 @@
+
+      - name: Install terraform-docs (binary)
+        run: |
+          curl -sSL https://github.com/terraform-docs/terraform-docs/releases/download/v0.19.0/terraform-docs-v0.19.0-linux-amd64.tar.gz -o terraform-docs.tar.gz
```
Let's keep all versions (including Python, Tofu and others) as env variables at the beginning of the workflow

rafaljanicki commented on terraform/buildspecs/martini-build-image.yaml:
```diff
@@ -0,0 +1,40 @@
+  pre_build:
+    commands:
+      - PARAMETER=$(aws ssm get-parameter --name "$PARAMETER_NAME" --with-decryption --query "Parameter.Value" --output text)
+      - export MARTINI_VERSION=$(echo "$PARAMETER" | jq -r '.martini_version')
```
Why is Martini version kept as a secret?

rafaljanicki commented on terraform/buildspecs/martini-upload-package.yaml:
```diff
@@ -0,0 +1,30 @@
+phases:
+  pre_build:
+    commands:
+      - PARAMETER=$(aws ssm get-parameter --name "$PARAMETER_NAME" --with-decryption --query "Parameter.Value" --output text)
```
Hm, fetching `PARAMETER_NAME` seems too broad by the name. Could you update the variable's name to be more descriptive?

rafaljanicki commented on terraform/modules/cloudwatch/main.tf:
```diff
@@ -0,0 +1,19 @@
+locals {
+  default_tags = {
```
Uhm, what's the point of this service tag?

rafaljanicki commented on terraform/modules/cloudwatch/main.tf:
```diff
@@ -0,0 +1,19 @@
+locals {
```
I wonder - if we're making everything module-based, why not use TF registry's modules instead?

rafaljanicki commented on terraform/pipelines/martini-build-image/main.tf:
```diff
@@ -0,0 +1,204 @@
+  ecr_repo_name           = "${local.resource_prefix}"
+  ssm_parameter_name      = "/martini/${local.environment}/${local.pipeline_name}"
+
+  common_tags = merge(
```
Let's make it provider's `default_tags`

rafaljanicki commented on terraform/pipelines/martini-build-image/README.md:
```diff
@@ -0,0 +1,103 @@
+
+## Requirements
+
+- Terraform 1.3.0+
```
You're using 1.6.6 in the pipeline, so that's confusing. Also, let's change to Tofu which is a free alternative

rafaljanicki commented on terraform/pipelines/martini-upload-package/main.tf:
```diff
@@ -0,0 +1,199 @@
+  codepipeline_role_name  = "${local.resource_prefix}-codepipeline-role"
+  ssm_parameter_name      = "/martini/${local.environment}/${local.pipeline_name}"
+
+  common_tags = merge(
```
Same as above

rafaljanicki commented on terraform/pipelines/martini-upload-package/README.md:
```diff
@@ -0,0 +1,122 @@
+
+## Requirements
+
+- Terraform 1.3.0+
```
Same

rafaljanicki commented on terraform/scripts/dockerfile:
```diff
@@ -1,5 +1,5 @@
 # Use the toroio/martini-runtime:${MARTINI_VERSION} image as the base
```
Shouldn't the dockerfile be camel case?

jasonquinto committed:
Switched to registry modules and other fixes

jasonquinto committed:
Fixed precommit ci workflow

jasonquinto committed:
Changed martini version from securestring to plaintext

jasonquinto commented:
Hi @rafaljanicki,  I’ve applied all your suggestions and requested changes, including switching to registry modules, using Tofu, aligning version usage, and updating the README and env vars. Let me know if I missed anything.

rafaljanicki commented on terraform/pipelines/martini-build-image/main.tf:
```diff
@@ -0,0 +1,253 @@
+
+module "project_log_group" {
+  source  = "terraform-aws-modules/cloudwatch/aws//modules/log-group"
+  version = "5.7.2"
```
`~>`

rafaljanicki commented on terraform/pipelines/martini-build-image/main.tf:
```diff
@@ -0,0 +1,253 @@
+
+module "pipeline_log_group" {
+  source  = "terraform-aws-modules/cloudwatch/aws//modules/log-group"
+  version = "5.7.2"
```
`~>`

rafaljanicki commented on terraform/pipelines/martini-build-image/main.tf:
```diff
@@ -0,0 +1,253 @@
+  # checkov:skip=CKV_AWS_300: Abort multipart uploads configured via lifecycle_rule
+
+  source  = "terraform-aws-modules/s3-bucket/aws"
+  version = "5.9.0"
```
`~>`

rafaljanicki commented on terraform/pipelines/martini-build-image/main.tf:
```diff
@@ -0,0 +1,253 @@
+}
+
+module "artifact_bucket" {
+
```
Empty space

rafaljanicki commented on terraform/pipelines/martini-build-image/main.tf:
```diff
@@ -0,0 +1,253 @@
+
+module "ecr" {
+  source  = "terraform-aws-modules/ecr/aws"
+  version = "3.1.0"
```
`~>`

rafaljanicki commented on terraform/pipelines/martini-build-image/main.tf:
```diff
@@ -0,0 +1,253 @@
+}
+
+module "build_image_parameter" {
+
```
empty space

rafaljanicki commented on terraform/pipelines/martini-build-image/main.tf:
```diff
@@ -0,0 +1,253 @@
+  # checkov:skip=CKV2_AWS_34
+
+  source  = "terraform-aws-modules/ssm-parameter/aws"
+  version = "2.0.1"
```
`~>`

rafaljanicki commented on terraform/pipelines/martini-upload-package/main.tf:
```diff
@@ -0,0 +1,266 @@
+
+module "project_log_group" {
+  source  = "terraform-aws-modules/cloudwatch/aws//modules/log-group"
+  version = "5.7.2"
```
`~>` (and the others, I'll stop making these comments)

rafaljanicki commented on terraform/pipelines/martini-upload-package/main.tf:
```diff
@@ -0,0 +1,266 @@
+
+module "artifact_bucket" {
+
+  # checkov:skip=CKV_AWS_21: Versioning explicitly enabled via module configuration
```
Does Checkov report on that?

jasonquinto replied:
Yes, even though I have versioning enabled.

rafaljanicki commented on .checkov.yaml:
```diff
@@ -0,0 +1,26 @@
+
+skip-check:
+  - CKV_TF_1            # Tofu/Terraform lockfile handling
+  - CKV_AWS_18          # S3 access logging not required for artifact bucket
```
That shouldn't be a global skip-check

rafaljanicki replied:
Same for the others here

jasonquinto committed:
Switch to version constraints and suppresed checkov policies on module level

jasonquinto commented:
Hi @rafaljanicki, I switched from pinned versions to version constraints for the registry modules, suppressed the Checkov policies at the module level instead of globally, and removed the empty spaces

rafaljanicki commented on terraform/pipelines/martini-build-image/main.tf:
```diff
@@ -0,0 +1,262 @@
+}
+
+module "build_image_parameter" {
+  # checkov:skip=CKV2_AWS_34
```
Add explanation why skipping

rafaljanicki commented on terraform/pipelines/martini-upload-package/main.tf:
```diff
@@ -0,0 +1,275 @@
+}
+
+module "build_image_parameter" {
+  # checkov:skip=CKV2_AWS_34
```
Add explanation why skipping

jasonquinto committed:
Added explanation for CKV2_AWS_34
